src/TracingCollector.ts

```diff
@@ -0,0 +1,117 @@
+import { Span, SpanBatch, SpanClient } from "@newrelic/telemetry-sdk/dist/src/telemetry/spans";
+import { Spec, TelemetrySpecMetadata } from "./types";
+
+const ERROR_FAIL_TO_EXTRACT_JSON = "Error: Invalid JSON parsing;";
+
+export default class TracingCollector {
+  serviceName: string;
+  specList: TelemetrySpecMetadata[] = [];
+  spanClient: SpanClient;
+
+  constructor(apiKey: string, serviceName: string) {
+    this.serviceName = serviceName;
+    this.spanClient = new SpanClient({
+      apiKey,
+    });
+  }
+
+  extractFullMessage(spec: any) {
+    try {
+      let fullMessage = null;
+      if (spec && spec.data && spec.data.msg) {
+          fullMessage = JSON.stringify(spec.data.msg);
+      }
+      return fullMessage
+    } catch (error) {
+      return ERROR_FAIL_TO_EXTRACT_JSON;
+    }
+  };
+
+  _extractFromSpec(spec: Spec, event: 'outward' | 'inward') {
+    const metadata: TelemetrySpecMetadata = {
+      id: spec.data.meta.id,
+      tx_id: spec.data.meta.tx,
+      mi_id: spec.data.meta.mi,
+      fullMessage: this.extractFullMessage(spec)
+    }
+    if (spec.ctx.actdef) {
+      metadata.plugin_name = spec.ctx.actdef.plugin_fullname;
+      metadata.pattern = spec.ctx.actdef.pattern;
+    }
+  
+    if (event === 'outward') {
+      metadata.duration = spec.ctx.duration;
+      metadata.endTime = spec.data.meta.end;
+      metadata.res = spec.data.res;
+      metadata.manualEndTime = Date.now();
+    }
+    if (event === 'inward' && !metadata.startTime) {
+      metadata.startTime = spec.data.meta.start;
+      metadata.manualStartTime = Date.now();
+    }
+    return metadata;
+  };
+
+  async _updateSpecList(specMetadata: TelemetrySpecMetadata) {
+    const spec = this.specList.find((s) => s.mi_id === specMetadata.mi_id);
+    if (spec) {
+      Object.assign(spec, specMetadata);
+      if (spec.manualEndTime && spec.manualStartTime && !spec.dispatched) {
+        try {
+          await this.sendTracing(spec);
+          spec.dispatched = true;
+          this._clearQueue();
+        } catch (error) {
+          console.log(error);
```

Comment:
replace all console.log with seneca.log.error|warn as appropriate
and remove all try catch not needed

NOTE: if just throwing the error, no need to log or catch